site-manage/site-manage-impl/impl/src/java/org/sakaiproject/sitemanage/impl/SiteManageServiceImpl.java

```diff
@@ -482,7 +493,48 @@ public void importToolsIntoSite(Site site, List<String> toolIds, Map<String, Lis
                     updateEntityReferences(toolId, toSiteId, transversalMap, site);
                 }
             }
+
+            // Handle the Context.id.history
+            mergeContextIdHistory(siteIds, site);
+        }
+    }
+
+    /**
+     * Compute the Context.id.history for the new site and insert it
+     *
+     * @param siteIds  a list of site ids to merge into the Context.id.history
+     * @param site       the site to save
+     */
+    private void mergeContextIdHistory(List<String> siteIds, Site site) {
+        List<String> new_list = new ArrayList<String>();
+        for(String fromSiteId : siteIds) {
+            try {
+                Site fromSite = siteService.getSite(fromSiteId);
+                ResourceProperties rp = fromSite.getProperties();
+                String old_id_history = rp.getProperty(CONTEXT_ID_HISTORY);
+                if ( StringUtils.isBlank(old_id_history) ) old_id_history = "";
+                List<String> old_id_list = Arrays.asList(old_id_history.split(","));
+
+                // Pull in the old ids.
+                for ( String old_id : old_id_list ) {
+                    if ( StringUtils.isBlank(old_id) ) continue;
+                    if ( ! new_list.contains(old_id) ) new_list.add(old_id);
```

Comment:
if a Set were used this would be simpler and safer, i.e. most of this can be reduced to one line
```new HashSet<>(Arrays.asList(oldIdHistory.split(","))).addAll(siteIds);```

Reply:
I think I want a LinkedHashSet because I want the order to be maintained.  I will give this a try.  Sets do remove a bunch of error-prone if statements *and* LinkedHashSet gives me the order I grave.  I am not a fan of super long single magic lines of a bunch of cascading calls that can harbor an NPE if the data is in a slightly different format than expected.